File: python/tests/arithmetic.py
import time, json, tracemalloc, gc, math

class ArithmeticTest:
    name = "Arithmetic"

    def __init__(self, n=10_000_000):
        self.n = n

    def workload(self):
        s_int = 0
        s_float = 0.0
        for i in range(1, self.n + 1):
            s_int += i * (i & 1)
            s_float += math.sin(i) * math.sqrt(i)
        return s_int, s_float

    def run_once(self):
        gc.collect()
        tracemalloc.start()
        t0 = time.perf_counter()

        s_int, s_float = self.workload()
        t1 = time.perf_counter()

        current, peak = tracemalloc.get_traced_memory()
        tracemalloc.stop()

        duration = t1 - t0

        return {
            "time_s": duration,
            "ops": self.n,
            "ops_per_sec": self.n / duration if duration > 0 else float("inf"),
            "mem_peak_bytes": peak,
            "acc_int": s_int,
            "acc_float": s_float
        }

    def run(self, runs=5, iterations=None):
        results = [self.run_once() for _ in range(runs)]

        times = sorted(r["time_s"] for r in results)
        ops = sorted(r["ops_per_sec"] for r in results)
        peaks = sorted(r["mem_peak_bytes"] for r in results)
        mid = len(results) // 2

        return {
            "name": self.name,
            "runs": runs,
            "median_time_s": times[mid],
            "median_ops_per_sec": ops[mid],
            "median_peak_mem_bytes": peaks[mid],
            "raw": results
        }

if __name__ == "__main__":
    result = ArithmeticTest(10_000_000).run()
    with open("results_arithmetic_python.json", "w") as f:
        json.dump(result, f, indent=2)
    print(json.dumps(result, indent=2))

File: ruby/tests/arithmetic.rb
require "json"
require "time"

class ArithmeticTest
  attr_reader :name

  def initialize(ops_per_iter = 10_000_000)
    @name = "Arithmetic"
    @ops_per_iter = ops_per_iter
  end

  def workload
    s_int = 0
    s_float = 0.0
    (1..999).each do |i|
      s_int += i * (i & 1)
      s_float += Math.sin(i) * Math.sqrt(i)
    end
    [s_int, s_float]
  end

  def run_once(iterations)
    reps = [1, iterations / 1000].max

    t0 = Process.clock_gettime(Process::CLOCK_MONOTONIC)
    acc_int = 0
    acc_float = 0.0

    reps.times do
      i, f = workload
      acc_int += i
      acc_float += f
    end

    duration = Process.clock_gettime(Process::CLOCK_MONOTONIC) - t0
    ops_done = reps * 1000
    ops_per_sec = ops_done / (duration > 0 ? duration : 1e-9)

    {
      "time_s" => duration,
      "ops" => ops_done,
      "ops_per_sec" => ops_per_sec,
      "acc_int" => acc_int,
      "acc_float" => acc_float,
      "gc_stat_after" => GC.stat
    }
  rescue Interrupt
    puts "\n[!] Arithmetic test interrupted — exiting safely."
    exit
  end

  def run(runs = 5, iterations = nil)
    iterations ||= @ops_per_iter
    results = []

    runs.times do
      results << run_once(iterations)
      GC.start
    end

    times = results.map { |r| r["time_s"] }.compact.sort
    ops = results.map { |r| r["ops_per_sec"] }.compact.sort

    median_time = times[times.length / 2]
    median_ops = ops[ops.length / 2]

    {
      "name" => @name,
      "runs" => runs,
      "median_time_s" => median_time,
      "median_ops_per_sec" => median_ops,
      "raw" => results
    }
  end
end

if __FILE__ == $0
  begin
    test = ArithmeticTest.new
    result = test.run
    File.write("results_arithmetic_ruby.json", JSON.pretty_generate(result))
    puts JSON.pretty_generate(result)
  rescue Interrupt
    puts "\n[!] Benchmark manually interrupted — partial data not saved."
  end
end
